gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -464,3 +469,179 @@ def _get_variant_calls(self, record, formats):
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+    # These members will be properly initiated in _extract_header_fields().
+    self._is_info_repeated = {}
+    self._is_format_repeated = {}
+    # This is a temporary solution until from_string will be fixed.
+    self._temp_local_file = self._write_to_local_file(file_name)
+
+  def _write_to_local_file(self, remote_file_name):
+    (temp_file, temp_file_name) = tempfile.mkstemp(text=True)
+    with FileSystems.open(remote_file_name) as f:
+      while True:
+        line = f.readline()
+        if line:
+          os.write(temp_file, line)
+        else:
+          os.close(temp_file)
+          break
+    return temp_file_name
+
+  def _store_to_temp_local_file(self, header_lines):
+    (temp_file, temp_file_name) = tempfile.mkstemp(text=True)
+    for line in header_lines:
+      os.write(temp_file, line)
+    os.close(temp_file)
+    return temp_file_name
+
+  def _init_with_header(self, header_lines):
+    # This header line is needed by Nucleus.
+    header_lines = ['##fileformat=VCFv4.2'] + header_lines
+    try:
+      # This is a temporary solution until from_string will be fixed.
+      self._vcf_reader = nucleus.io.vcf.VcfReader(
+          self._temp_local_file, use_index=False)
+
+    except SyntaxError as e:
+      raise ValueError(
+          'Invalid VCF header in %s: %s' % (self._file_name, str(e)))
+    self._extract_header_fields()
+
+  def _extract_header_fields(self):
+    header = self._vcf_reader.header
+    for info in header.infos:
+      if info.number in ('0', '1'):
+        self._is_info_repeated[info.id] = False
+      else:
+        self._is_info_repeated[info.id] = True
+
+    for format_info in header.formats:
+      if format_info.number in ('0', '1'):
+        self._is_format_repeated[format_info.id] = False
+      else:
+        self._is_format_repeated[format_info.id] = True
+
+  def _get_variant(self, data_line):
+    try:
+      # This is a temporary solution until from_string will be fixed.
+      record = next(self._vcf_reader)
+      return self._convert_to_variant(record)
+    except (LookupError, ValueError) as e:
+      logging.warning('VCF record read failed in %s for line %s: %s',
+                      self._file_name, data_line, str(e))
+      return MalformedVcfRecord(self._file_name, data_line, str(e))
+
+  def _convert_to_variant(self, record):
+    # type: (nucleus_proto.Variant) -> Variant
+    return Variant(
+        reference_name=record.reference_name,
+        start=record.start,
+        end=record.end,
+        reference_bases=(record.reference_bases
+                         if record.reference_bases != MISSING_FIELD_VALUE
+                         else None),
+        alternate_bases=(
+            map(str, record.alternate_bases) if record.alternate_bases else []),
+        names=map(str, record.names) if record.names else [],
+        quality=record.quality,
```

Comment:
I think you should check for has_quality first, as otherwise, this would always return 0 if unspecified (we should set it to null if unspecified).

Reply:
Right; it seems like they use `-1` instead of null for quality field. We should eventually revisit this and special case on that. OK to leave as TODO for this PR.